File: scala/problem16.scala
/*
https://projecteuler.net/problem=16

Power digit sum
*/

object problem16 {
  import scala.collection.mutable.ArrayBuffer

  def baselineAnswer: Long = {
    BigInt(2).pow(1000).toString.map(_.asDigit.toLong).sum
  }

  def optimizedAnswer: Long = {
    val digits = ArrayBuffer[Int](1)
    var exp = 0
    while (exp < 1000) {
      var carry = 0
      var i = 0
      while (i < digits.length) {
        val value = digits(i) * 2 + carry
        digits(i) = value % 10
        carry = value / 10
        i += 1
      }
      while (carry > 0) {
        digits += (carry % 10)
        carry /= 10
      }
      exp += 1
    }

    digits.foldLeft(0L)(_ + _.toLong)
  }

  case class Result(value: Long, timeNs: Long)

  def answer: Long = {
    optimizedAnswer
  }

  def eulerBaseline(): Result = {
    val start = System.nanoTime()
    val value = baselineAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def euler(): Result = {
    val start = System.nanoTime()
    val value = optimizedAnswer
    val end = System.nanoTime()
    Result(value, end - start)
  }

  def main(args: Array[String]): Unit = {
    println("Project Euler Problem 16")

    val baseline = eulerBaseline()
    val optimized = euler()

    println(s"Result (baseline): ${baseline.value}")
    println(s"Time (baseline): ${baseline.timeNs} ns")
    println(s"Result (optimized): ${optimized.value}")
    println(s"Time (optimized): ${optimized.timeNs} ns")

    if (optimized.timeNs > 0) {
      val speedup = baseline.timeNs.toDouble / optimized.timeNs.toDouble
      println(s"Speedup: ${speedup}x")
    }
  }
}

File: cpp/problem16.cpp
/*
https://projecteuler.net/problem=16

Power digit sum

What is the sum of the digits of the number 2^1000?
*/

#include <chrono>
#include <iostream>
#include <string>
#include <vector>
using namespace std;

struct result {
    long long value;
    long long time_ns;
};

int digit_sum_base10(std::vector<int> d) {
    int sum = 0;
    for (int x : d) sum += x;
    return sum;
}

long long evaluate_baseline() {
    std::vector<int> digits(1, 1);

    for (int i = 0; i < 1000; ++i) {
        int carry = 0;
        for (size_t j = 0; j < digits.size(); ++j) {
            int v = digits[j] * 2 + carry;
            digits[j] = v % 10;
            carry = v / 10;
        }
        while (carry > 0) {
            digits.push_back(carry % 10);
            carry /= 10;
        }
    }

    return digit_sum_base10(digits);
}

long long evaluate_optimized() {
    const int BASE = 1000000000;
    std::vector<int> chunks(1, 1);

    for (int i = 0; i < 1000; ++i) {
        long long carry = 0;
        for (size_t j = 0; j < chunks.size(); ++j) {
            long long v = 2LL * chunks[j] + carry;
            chunks[j] = static_cast<int>(v % BASE);
            carry = v / BASE;
        }
        if (carry > 0) {
            chunks.push_back(static_cast<int>(carry));
        }
    }

    long long sum = 0;
    for (size_t i = 0; i < chunks.size(); ++i) {
        int part = chunks[i];
        int width = (i + 1 == chunks.size()) ? 0 : 9;
        int count = 0;

        if (part == 0) {
            count = 1;
        }

        while (part > 0) {
            sum += part % 10;
            part /= 10;
            ++count;
        }

        while (count < width) {
            ++count;
        }
    }

    return sum;
}

result euler() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_optimized();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

result euler_baseline() {
    const auto start = std::chrono::steady_clock::now();
    long long value = evaluate_baseline();
    const auto end = std::chrono::steady_clock::now();
    long long elapsed = std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count();
    return {value, elapsed};
}

int main() {
    std::cout << "Project Euler Problem 16" << std::endl;

    result baseline = euler_baseline();
    result optimized = euler();

    std::cout << "Result (baseline): " << baseline.value << std::endl;
    std::cout << "Time (baseline): " << baseline.time_ns << " ns" << std::endl;
    std::cout << "Result (optimized): " << optimized.value << std::endl;
    std::cout << "Time (optimized): " << optimized.time_ns << " ns" << std::endl;

    if (optimized.time_ns > 0) {
        double speedup = static_cast<double>(baseline.time_ns) / static_cast<double>(optimized.time_ns);
        std::cout << "Speedup: " << speedup << "x" << std::endl;
    }

    return 0;
}
